File: 프로그래머스/2/68645. 삼각 달팽이/삼각 달팽이.py
def check(i, j, n):
    if (i < 0 or j < 0 or j > i or i >= n):
        return False
    return True

def solution(n):
    arr = []
    answer = []
    dir = [[1, 0], [0, 1], [-1, -1]]
    num = n * (n + 1) / 2
    for i in range(n):
        arr.append([0] * (i + 1))
    x = 2
    d = 0
    [i, j] = [0, 0]
    arr[0][0] = 1
    while x <= num:
        [ni, nj] = [i + dir[d][0], j + dir[d][1]]
        if (check(ni, nj, n) == False or arr[ni][nj] != 0):
            d += 1
            d %= 3
            [ni, nj] = [i + dir[d][0], j + dir[d][1]]
        arr[ni][nj] = x
        [i, j] = [ni, nj]
        x += 1
    for a in arr:
        answer += a
    return answer

File: 프로그래머스/2/68645. 삼각 달팽이/삼각 달팽이.js
function solution(n) {
    let arr = new Array(n).fill(0);
    for (let i = 0; i < n; i++)
        arr[i] = new Array(i + 1).fill(0);
    arr[0][0] = 1;
    let answer = [];
    let x = 2;
    let num = n * (n + 1) / 2;
    let i = 0; let j = 0; let d = 0;
    let dir = [[1, 0], [0, 1], [-1, -1]];
    const check = (i, j) => {
        if (i < 0 || j < 0 || i < j || i >= n) return false;
        return true;
    }
    
    while (x <= num) {
        let ni = dir[d][0] + i;
        let nj = dir[d][1] + j;
        if (!check(ni, nj) || arr[ni][nj]) {
            d++; d %= 3;
            ni = dir[d][0] + i; nj = dir[d][1] + j;
        }
        arr[ni][nj] = x;
        x++; i = ni; j = nj;
    }
    arr.forEach(e => answer = [...answer, ...e]);
    return answer;
}